llvm/test/CodeGen/RISCV/rvv/fixed-vectors-insert.ll

```diff
@@ -550,9 +550,9 @@ define void @insertelt_c6_v8i64_0_add(ptr %x, ptr %y) {
 ; CHECK-NEXT:    vsetivli zero, 8, e64, m4, ta, ma
 ; CHECK-NEXT:    vle64.v v8, (a0)
 ; CHECK-NEXT:    li a2, 6
-; CHECK-NEXT:    vsetvli zero, zero, e64, m4, tu, ma
+; CHECK-NEXT:    vsetivli zero, 8, e64, m1, tu, ma
```

Comment:
Continuing from our chat offline... Ah, here's the problem I hit when I tried this before.

8 is not a valid VL for m1.  Trying to set VL=8 will result in a VL=2 on zvl126b hardware.  This is correct for the vmv.s.x, but forces two pairs of VL toggles in this bit of assembly.  (We'd previously had two VTYPE toggles, but VTYPE is cheaper than VL.)

This isn't a correctness issue, but it does raise a question of profitability.  Is it worth two VL toggles to use a narrower LMUL here?  I'm not sure it is.  

Alternatively, can we reasonably strengthen InsertVSETVLI to undo the damage here?  Maybe we should allow the m4 to slide over the vmv.s.x if doing so eliminates the VL toggle?

Reply:
Isn't VL is already 8 here? Maybe this is just an issue with InsertVSETVLI not using the `vsetvli zero, zero` form?

Reply:
Did a bit of digging, InsertVSETVLI has to set VL again because VLMAX has changed with LMUL. Allowing the m4 to slide over the vmv.s.x sounds like a good path to take here

Reply:
Normally the m4 would already slide over the vmv.s.x here because we explicitly mark LMUL and VL as not demanded (we just check that VL has the same zero-ness). I think the reason as to why this the vsetvlis aren't being merged is because of the tail policy switch. 

[This old patch](https://reviews.llvm.org/D156319) that relaxed TA -> TU eliminates the toggle for this test (but it's overall not profitable for OoO processors): https://reviews.llvm.org/differential/changeset/?ref=4362822

For this specific test case though, should we not be doing this as a vmv.v.i with VL=1 anyway?

Reply:
I think there's a legality issue here.  Since the vmv.s.x has a register class which only requires m1, it could chose an unaligned (for m4) register.  This would result in an illegal instruction if we allow the m4 to forward over the m1 vmv.s.x.  We'd have to change the register class constraints to do this legally.

Reply:
I think I'm missing something here. Are you saying that we don't currently allow a m1 vmv.s.x to be turned into a m4 vmv.s.x?

If we construct this example below where the vadd.vi has m4 (%1:vrm4), and the vmv.s.x has m1 (%3:vr)

```
---
name: lmul_merge
body:        |
  bb.0:
    liveins: $v0, $x1, $x2
    %vl:gprnox0 = COPY $x1

    %0:vrm4 = COPY $v0
    %pt0:vrm4 = IMPLICIT_DEF
    %1:vrm4 = PseudoVADD_VI_M4 %pt0, %0, 1, %vl, 6, 0
    
    %2:gpr = COPY $x2
    %pt1:vr = IMPLICIT_DEF
    %3:vr = PseudoVMV_S_X_M1 %pt1, %2, %vl, 6
    PseudoRET
```

Running `llc -o - -run-pass=riscv-insert-vsetvli -mtriple=riscv64 -mattr=+v -verify-machineinstrs`, it looks like we do relax the LMUL for vmv.s.x:

```
body:             |
  bb.0:
    liveins: $v0, $x1, $x2
  
    %vl:gprnox0 = COPY $x1
    %1:vrm4 = COPY $v0
    %pt0:vrm4 = IMPLICIT_DEF
    dead $x0 = PseudoVSETVLI %vl, 218 /* e64, m4, ta, ma */, implicit-def $vl, implicit-def $vtype
    %3:vrm4 = PseudoVADD_VI_M4 %pt0, %1, 1, $noreg, 6 /* e64 */, 0 /* tu, mu */, implicit $vl, implicit $vtype
    %4:gpr = COPY $x2
    %pt1:vr = IMPLICIT_DEF
    %6:vr = PseudoVMV_S_X_M1 %pt1, %4, $noreg, 6 /* e64 */, implicit $vl, implicit $vtype
    PseudoRET
```

I see what you mean about the legality though: Does the register allocator here know not to allocate v1/v2/v3 etc for %6, given its vtype will be M4?

Reply:
> I see what you mean about the legality though: Does the register allocator here know not to allocate v1/v2/v3 etc for %6, given its vtype will be M4?

I think I'm wrong here and I sent you on a wild goose chase.  The specification says for vmv.s.x that "The instructions ignore LMUL and vector register groups.".  I think that means these are unconditionally m1 (from an encoding legality perspective) even if LMUL is set to m4.